### PR TITLE
Add getting recent docker events

### PR DESCRIPTION
**Pull request by ggodlewski**
Allow to pull recent recent containers events

### CODE REVIEW

**Comment by hacfi** [comment]
:+1: LGTM

File: src/Docker/Docker.php [hunk 1]
```diff
@@ -177,4 +176,5 @@ public function commit(Container $container, $config = [])
 
         return $image;
     }
+
```
Comment (hacfi) [inline src/Docker/Docker.php]: That’s not really necessary but whatever.

**Comment by ubermuda** [comment]
Is the `ContainerManager` really the right place for this? If I understand correctly, this has nothing to do specifically with containers, so maybe it should be in `Docker`?

Coding standards: we usually put a white line before and after structure blocks like `for`s and `if`s, except before a block that begins another block. For example, the `splitEvents` method would look like this:

```php
        // ...
        $currentString = '';

        for ($i = 0; $i < strlen($contents); $i++) {
            if ($contents{$i} == '{') {
                $numOfBrackets++;
            }

            if ($contents{$i} == '}') {
                $numOfBrackets--;
            }

            $currentString .= $contents{$i};

            if ($numOfBrackets == 0 ) {
                if (!empty($currentString)) {
                    $retVal[] = new Event(json_decode($currentString, true));
                }

                $currentString = '';
            }
        }
        // ...
```

Also, +1 for the documentation, and this will also needs tests :)

**Comment by ggodlewski** [comment]
I moved it into ContainerManager because those are container events. Maybe some new EventMenager? Tell me where you want it and I'll correct it and add docs.

**Comment by ubermuda** [comment]
We are talking about http://docs.docker.com/reference/api/docker_remote_api_v1.18/#monitor-dockers-events, right?

**Comment by ggodlewski** [comment]
Yes.

**Comment by ubermuda** [comment]
Well that page says `/events` can report containers events AND images events

**Comment by ubermuda** [comment]
Then it can also report image events:

```
and Docker images will report:

    untag, delete
```

So yeah, maybe an `EventManager` would make sense, maybe something based on https://github.com/symfony/EventDispatcher so that we can easily register listeners to these events, what do you think?

**Comment by ggodlewski** [comment]
At the moment it uses polling. If you want to dispatch events it should use streaming. It should open /events and asynchronously react to received data converting it to events.

I might be wrong (don't know guzzle so good) but it seems to require guzzle5 with its asynchronous capabilities.

**Comment by ubermuda** [comment]
Right, I guess both polling and streaming can have their place in docker-php.

In the end, because `/events` can report image events, it shouldn't be in `ContainerManager`, but there might not be a need for an EventManager right now (you're right about the streaming, and we're already talking about switching to egeloen/ivory-http-adapter which would bring in async + psr-7 compliance, so this part should wait) so maybe you can move it to `Docker\Docker` and that would be fine for now, what do you think?

**Comment by ggodlewski** [comment]
Hey, I put it into new place. It's not a lot of code but when async is added it will more. I don't want to change api then.
And there is a unit test now.

File: src/Docker/Manager/EventManager.php [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+     * Split json events text to array
+     *
+     * @param $contents
+     * @return array
+     */
+    public function splitEvents($contents)
+    {
+        $retVal = [];
+        $numOfBrackets = 0;
+        $currentString = '';
```
Comment (ubermuda) [inline src/Docker/Manager/EventManager.php]: Empty line after this

File: src/Docker/Manager/EventManager.php [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        for ($i = 0; $i < strlen($contents); $i++) {
+            if ($contents{$i} == '{') {
+                $numOfBrackets++;
+            }
```
Comment (ubermuda) [inline src/Docker/Manager/EventManager.php]: Empty line after this

File: src/Docker/Manager/EventManager.php [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            if ($contents{$i} == '}') {
+                $numOfBrackets--;
+            }
```
Comment (ubermuda) [inline src/Docker/Manager/EventManager.php]: Empty line after this

File: src/Docker/Manager/EventManager.php [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            $currentString .= $contents{$i};
+            if ($numOfBrackets == 0 ) {
```
Comment (ubermuda) [inline src/Docker/Manager/EventManager.php]: Empty line **before** this

File: src/Docker/Manager/EventManager.php [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+                if (!empty($currentString)) {
+                    $retVal[] = new Event(json_decode($currentString, true));
+                }
```
Comment (ubermuda) [inline src/Docker/Manager/EventManager.php]: Empty line after this

File: src/Docker/Tests/Manager/EventManagerTest.php [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        $containerManager->create($container);
+
+        $eventManager = $this->getDocker()->getEventManager();
+        $events = $eventManager->getEvents($since, time());
+
+        $this->assertNotEmpty($container->getId());
+        $this->assertThat($events, new PHPUnit_Framework_Constraint_Callback(function ($events) use($container) {
+            foreach ($events as $event) {
+                if ($event->getStatus() == 'create' && $event->getId() == $container->getId()) return true;
+            }
```
Comment (ubermuda) [inline src/Docker/Tests/Manager/EventManagerTest.php]: Empty line after this

File: src/Docker/Tests/Manager/EventManagerTest.php [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+        $containerManager = $this->getDocker()->getContainerManager();
+        $containerManager->create($container);
+
+        $eventManager = $this->getDocker()->getEventManager();
+        $events = $eventManager->getEvents($since, time());
+
+        $this->assertNotEmpty($container->getId());
+        $this->assertThat($events, new PHPUnit_Framework_Constraint_Callback(function ($events) use($container) {
+            foreach ($events as $event) {
+                if ($event->getStatus() == 'create' && $event->getId() == $container->getId()) return true;
```
Comment (ubermuda) [inline src/Docker/Tests/Manager/EventManagerTest.php]: Please do not use control structures without `{}`s

**Comment by ubermuda** [comment]
Ok that's great, ready to merge once the coding style issues are fixed!

**Comment by ubermuda** [comment]
poke @ggodlewski, are you going to fix the CS issues? :)

**Comment by schmunk42** [comment]
Why was this closed without merge? Are these changes in the code or in another PR?

**Comment by ggodlewski** [comment]
As you see there were some empty lines missing and instead of just putting them he spent more time on writing comments.
I have a life. I prefer to use and contribute to another project.

**Comment by schmunk42** [comment]
@ggodlewski Thank you for the info, sad to hear the PR failed just because of code-formatting.

**Comment by ubermuda** [comment]
@ggodlewski sorry that you feel this way, but code formatting is important. If you do not wish to spend the time to respect the work of others by complying to simple rules, that is your right and I hope you find another project to use and contribute to that will suit your needs and philosophy.

In the meantime, I would just like to point that being sarcastic and/or ironic is not necessary, and implying that your time is more valuable than that of other people here is, simply put, rude.

**Pull request closed without merging**